structuralcodes/codes/ec2_2004/_section_7_3_crack_control.py

```diff
@@ -0,0 +1,933 @@
+        ValueError: if any of phi1 or phi2 is less than 0
+        TypeError: if any of n1 or n2 is not an integer
+    """
+    if n1 < 0:
+        raise ValueError(f'n1={n1} cannot be less than 0')
+    if not isinstance(n1, int):
+        raise TypeError(f'n1={n1} needs to be an integer value')
+    if n2 < 0:
+        raise ValueError(f'n2={n2} cannot be less than 0')
+    if not isinstance(n2, int):
+        raise TypeError(f'n2={n2} needs to be an integer value')
+    if phi1 < 0:
+        raise ValueError(f'phi1={phi1} cannot be less than 0')
+    if phi2 < 0:
+        raise ValueError(f'phi2={phi2} cannot be less than 0')
+
+    a = n1 * phi1**2 + n2 * phi2**2
+    b = n1 * phi1 + n2 * phi2
+    return a / b
+
+
+def k1(bond_type: str) -> float:
+    """Get the k1 coefficient which takes account of the bond properties
+    of the bounded reinforcement
+
+    EUROCODE 2 1992-1-1:2004, Eq. (7.11-k1)
+
+    Args:
+        bond_type (str): the bond property of the reinforcement.
+        Possible values:
+            - 'bond': for high bond bars
+            - 'plane': for bars with an effectively plain surface (e.g.
+            prestressing tendons)
+
+    Returns:
+        (float): value of the k1 coefficient
+
+    Raises:
+        ValueError: if bond_type is neither 'bond' nor 'plane'
+        TypeError: if bond_type is not an str
+    """
+    if not isinstance(bond_type, str):
+        raise TypeError(f'bond_type={bond_type} is not an str')
+
+    bond_type = bond_type.lower().strip()
+    if bond_type != 'bond' and bond_type != 'plane':
+        raise ValueError(
+            f'bond_type={bond_type} can only have "bond" or "plane" as values'
+        )
+
+    return 0.8 if bond_type == 'bond' else 1.6
+
+
+def k2(epsilon_r: float) -> float:
+    """Computes a coefficient which takes into account of the
+    distribution of strain:
+
+    EUROCODE 2 1992-1-1:2004, Eq. (7.13)
+
+    Args:
+        epsilon_r (float): ratio epsilon_2/epsilon_1 where epsilon_1 is
+            thre greater and epsilon_2 is the lesser strain at the boundaries
+            of the section considererd, assessed on the basis of a cracked
+            section. epsilon_r=0 for bending and epsilon_r=1 for pure tension.
+
+    Returns:
+        float: the k2 coefficient value.
+
+    Raises:
+        ValueError: if epsilon_r is not between 0 and 1.
+    """
+    if epsilon_r < 0 or epsilon_r > 1:
+        raise ValueError(f'epsilon_r={epsilon_r} must be between 0 and 1')
+
+    return (1 + epsilon_r) / 2
+
+
+def k3():
+    """Returns the k3 coefficient for computing sr_max
+
+    Returns:
+        float: value for the coefficient
+    """
+    return 3.4
+
+
+def k4():
+    """Returns the k4 coefficient for computing sr_max
+
+    Returns:
+        float: value for the coefficient
+    """
+    return 0.425
+
+
+def sr_max_close(
+    c: float,
+    phi: float,
+    _rho_p_eff: float,
+    _k1: float,
+    _k2: float,
+    _k3: float,
+    _k4: float,
```

Comment:
Consider making the `k`s optional, e.g. `_k1: t.Optional[float] = None`, and use the relevant `k` function for computing it's value if no value is provided.

Reply:
Great idea! 

What do you think about adding the option of being None only to `_k3` and `_k4`? Parameters `_k1` and `_k2` require as input values `bond_type` and `epsilon_r` respectively to be computed.

I have also included the tests if these values are None.

Reply:
Good idea, keep `_k3` and `_k4` optional and the rest required 😃